File: java/PermutationsII.java
import java.util.*;

/**

 Given a collection of numbers that might contain duplicates, return all possible unique permutations.

 For example,
 [1,1,2] have the following unique permutations:
 [
 [1,1,2],
 [1,2,1],
 [2,1,1]
 ]

 */

public class PermutationsII {
    public List<List<Integer>> permuteUnique(int[] nums) {
        List<List<Integer>> result = new ArrayList<>();

        Arrays.sort(nums);
        int[] visited = new int[nums.length];


        helper(result, new ArrayList<Integer>(), nums, visited);


        return result;
    }

    public void helper(List<List<Integer>> result, List<Integer> cur, int[] nums, int[] visited) {

        if (cur.size() == nums.length) {

            result.add(new ArrayList<>(cur));

            return;
        }

        for (int i = 0; i < nums.length; i++) {

            if (visited[i] == 1 || i > 0 && nums[i] == nums[i - 1] && visited[i - 1] == 0) {
                continue;
            }

            cur.add(nums[i]);
            visited[i] = 1;

            helper(result, cur, nums, visited);

            cur.remove(cur.size() - 1);
            visited[i] = 0;
        }
    }
}

File: -Old/Stack/Permutation/permutations_ii.py
'''
Description
Given a list of numbers with duplicate number in it. Find all unique permutations.

Have you met this question in a real interview?
Example
For numbers [1,2,2] the unique permutations are:

[
  [1,2,2],
  [2,1,2],
  [2,2,1]
]
Challenge
Using recursion to do it is acceptable. If you can do it without recursion, that would be great!

'''

def permuteUnique(self, nums):
  if nums is None:
    return []

  result = []
  nums.sort()
  visited = [False] * len(nums)
  cur = []
  results = []
  self.dfs(nums, visited, cur, result)

  return result

def dfs(self, nums, visited, cur, result):
  if len(cur) == len(nums):
    result.append(list(cur))
    return

  for i in range(len(nums)):
    if visited[i]:
      continue

    if i != 0 and nums[i] == nums[i - 1] and not visited[i - 1]:
      continue

    cur.append(nums[i])
    visited[i] = True
    self.dfs(nums, visited, cur, result)
    visited[i] = False
    cur.pop()